### PR TITLE
CMakeList refactor

### PR DESCRIPTION
**Pull request by fgalan**
This PR tries to reduce the entropy in the CMakeList.txt system. Common libraries has been "abstracted" to the master CMakeList.txt to make the system more maintenable. In the process some unneeded libraries (e.g. ssl)  has been removed.

This has been validaded running `make debug` and `make unit_test` in the following distros:

* Debian 8.2
* Debian 7
* CentOS 6
* CentOS 7

Others distros not yet verified are "marked" in the files as pending.

@crbrox 
@kzangeli

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -45,17 +45,13 @@ execute_process(COMMAND            echo $USER
 execute_process(COMMAND            sh ../scripts/build/osDistro.sh
                 OUTPUT_VARIABLE    DISTRO)
 
-execute_process(COMMAND            sh ../scripts/rapidxmlVersion.sh
```
Comment (fgalan) [inline CMakeLists.txt]: Not sure in which moment in the past this wasn't needed anymore...

Comment (kzangeli) [inline CMakeLists.txt]: Probably in the first PR about removing XML ...

NTC

File: src/app/contextBroker/CMakeLists.txt [hunk 2]
```diff
@@ -27,81 +27,60 @@ SET (HEADERS
...
-  MESSAGE("contextBroker: Debian DISTRO: '${DISTRO}'")
-  TARGET_LINK_LIBRARIES(contextBroker ${STATIC_LIBS} -lmongoclient -lboost_thread -lboost_filesystem -lboost_system -lpthread -lboost_regex -lssl -lgcrypt -luuid)
-ELSE()
-  MESSAGE("contextBroker: Other DISTRO: '${DISTRO}'")
-  TARGET_LINK_LIBRARIES(contextBroker ${STATIC_LIBS} -lmongoclient -lboost_thread -lboost_filesystem -lboost_system -lpthread -luuid)
+  TARGET_LINK_LIBRARIES(contextBroker ${STATIC_LIBS} ${BOOST} ${DYNAMIC_LIBS})
+
+# The following ones are distros that haven't been checked since the refactoring in
+# PR #2062. They should be "promoted" as soon as we have positive feedback about
+# tehm or fixed (maybe based in the version of this file previous to
```
Comment (kzangeli) [inline src/app/contextBroker/CMakeLists.txt]: typo: tehm

File: src/app/contextBroker/CMakeLists.txt [hunk 3]
```diff
@@ -27,81 +27,60 @@ SET (HEADERS
...
+  TARGET_LINK_LIBRARIES(contextBroker ${STATIC_LIBS} ${BOOST} ${DYNAMIC_LIBS})
+ELSEIF((${DISTRO} STREQUAL "Ubuntu_14.04.2_LTS") OR (${DISTRO} STREQUAL "Ubuntu_14.04.3_LTS") OR (${DISTRO} STREQUAL "Ubuntu_14.04.4_LTS") OR (${DISTRO} MATCHES "Linux_Mint_.*"))    
+  TARGET_LINK_LIBRARIES(contextBroker ${STATIC_LIBS} ${BOOST} ${DYNAMIC_LIBS})
+ELSEIF(${DISTRO} STREQUAL "openSUSE_13.1")  
+  TARGET_LINK_LIBRARIES(contextBroker ${STATIC_LIBS} ${BOOST_MT} ${DYNAMIC_LIBS})
+ELSEIF(${DISTRO} STREQUAL "Fedora_20")  
+  TARGET_LINK_LIBRARIES(contextBroker ${STATIC_LIBS} ${BOOST} ${DYNAMIC_LIBS})
+
+# Finally this is a "best effort" option to try to build on distros that don't have an
+# specify switch
```
Comment (kzangeli) [inline src/app/contextBroker/CMakeLists.txt]: don't have a specific switch

File: src/app/contextBroker/CMakeLists.txt [hunk 4]
```diff
@@ -27,81 +27,60 @@ SET (HEADERS
...
+
+ELSE()  
+  TARGET_LINK_LIBRARIES(contextBroker ${STATIC_LIBS} -lmongoclient ${BOOST} ${DYNAMIC_LIBS})
 ENDIF()
```
Comment (kzangeli) [inline src/app/contextBroker/CMakeLists.txt]: Nice grouping of libs. Much easier to read now.

NTC

File: src/lib/ngsi/ContextAttribute.cpp [hunk 5]
```diff
@@ -770,8 +769,7 @@ std::string ContextAttribute::toJson(bool isLastElement, const std::string& rend
   //   Perhaps these two parameters (renderMode and notifyFormat) should be unified ...
   //
   if (rMode == RENDER_MODE_UNIQUE_VALUES)
-  {
-    uniqueValues = true;
+  {    
     rMode        = RENDER_MODE_VALUES;  // FIXME PR: is this correct?
   }
```
Comment (kzangeli) [inline src/lib/ngsi/ContextAttribute.cpp]: This seems to be from the previous PR, as it is still not in develop ... ?

Comment (fgalan) [inline src/lib/ngsi/ContextAttribute.cpp]: Yes. However, I understand that as long as the change is exactly the same, there wouldn't be any merging problem.

NTC

File: src/app/contextBroker/CMakeLists.txt (hunk 3 above)
Comment (fgalan) [inline src/app/contextBroker/CMakeLists.txt]: Fixed in 5da6de8

File: src/app/contextBroker/CMakeLists.txt (hunk 2 above)
Comment (fgalan) [inline src/app/contextBroker/CMakeLists.txt]: Fixed in 5da6de8

File: src/lib/ngsi/ContextAttribute.cpp (hunk 5 above)
Comment (kzangeli) [inline src/lib/ngsi/ContextAttribute.cpp]: Yes. No probs.

NTC

**Comment by kzangeli** [comment]
LGTM

**Comment by crbrox** [comment]
LGTM

**Pull request merged**